File: WebDriver-Series/WebDriver.Series.Tests/GridTestCases/ShipNameColumnAsserter.cs
using System;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using WebDriver.Series.Tests.Controls;
using WebDriver.Series.Tests.GridTestCases.Pages;

namespace WebDriver.Series.Tests.GridTestCases
{
    public class ShipNameColumnAsserter : GridColumnAsserter
    {
        public ShipNameColumnAsserter(IGridPage gridPage) : base(gridPage)
        {
        }

        public void ShipNameEqualToFilter()
        {
            this.GridPage.NavigateTo();
            var newItem = this.CreateNewItemInDb();

            this.GridPage.Grid.Filter(GridColumns.ShipName, Enums.FilterOperator.EqualTo, newItem.ShipName);
            this.WaitForGridToLoad(1, this.GridPage.Grid);
            var items = this.GridPage.Grid.GetItems<GridItem>();
            
            Assert.AreEqual(1, items.Count);
        }
        
        public void ShipNameContainsFilter()
        {
            this.GridPage.NavigateTo();
            string shipName = Guid.NewGuid().ToString();
            // Remove first and last letter
            shipName = shipName.TrimStart(shipName.First()).TrimEnd(shipName.Last());
            var newItem = this.CreateNewItemInDb(shipName);          

            this.GridPage.Grid.Filter(GridColumns.ShipName, Enums.FilterOperator.Contains, newItem.ShipName);
            this.WaitForGridToLoad(1, this.GridPage.Grid);
            var items = this.GridPage.Grid.GetItems<GridItem>();
            
            Assert.AreEqual(1, items.Count);
        }
        
        public void ShipNameEndsWithFilter()
        {
            this.GridPage.NavigateTo();

            // Remove first letter 
            string shipName = Guid.NewGuid().ToString();
            shipName = shipName.TrimStart(shipName.First());
            var newItem = this.CreateNewItemInDb(shipName);

            this.GridPage.Grid.Filter(GridColumns.ShipName, Enums.FilterOperator.EndsWith, newItem.ShipName);
            this.WaitForGridToLoad(1, this.GridPage.Grid);
            var items = this.GridPage.Grid.GetItems<GridItem>();
            
            Assert.AreEqual(1, items.Count);
        }
        
        public void ShipNameStartsWithFilter()
        {
            this.GridPage.NavigateTo();

            // Remove last letter
            string shipName = Guid.NewGuid().ToString();
            shipName = shipName.TrimEnd(shipName.Last());
            var newItem = this.CreateNewItemInDb(shipName);

            this.GridPage.Grid.Filter(GridColumns.ShipName, Enums.FilterOperator.StartsWith, newItem.ShipName);
            this.WaitForGridToLoad(1, this.GridPage.Grid);
            var items = this.GridPage.Grid.GetItems<GridItem>();
            
            Assert.AreEqual(1, items.Count);
        }
        
        public void ShipNameNotEqualToFilter()
        {
            this.GridPage.NavigateTo();
            
            // Apply combined filter. First filter by ID and than by ship name (not equal filter). 
            // After the first filter there is only one element when we apply the second we expect 0 elements.
            var newItem = this.CreateNewItemInDb();
            
            this.GridPage.Grid.Filter(
                new GridFilter(GridColumns.ShipName, Enums.FilterOperator.NotEqualTo, newItem.ShipName),
                new GridFilter(GridColumns.OrderID, Enums.FilterOperator.EqualTo, newItem.OrderId.ToString()));
            this.WaitForGridToLoad(0, this.GridPage.Grid);
            var items = this.GridPage.Grid.GetItems<GridItem>();
            
            Assert.AreEqual(0, items.Count);
        }
        
        public void ShipNameNotContainsFilter()
        {
            this.GridPage.NavigateTo();
          
            // Remove first and last letter
            string shipName = Guid.NewGuid().ToString();
            shipName = shipName.TrimStart(shipName.First()).TrimEnd(shipName.Last());
            var newItem = this.CreateNewItemInDb(shipName);

            // Apply combined filter. First filter by ID and than by ship name (not equal filter). 
            // After the first filter there is only one element when we apply the second we expect 0 elements.
            this.GridPage.Grid.Filter(
                new GridFilter(GridColumns.ShipName, Enums.FilterOperator.NotContains, newItem.ShipName),
                new GridFilter(GridColumns.OrderID, Enums.FilterOperator.EqualTo, newItem.OrderId.ToString()));
            this.WaitForGridToLoad(0, this.GridPage.Grid);
            var items = this.GridPage.Grid.GetItems<GridItem>();
            
            Assert.AreEqual(0, items.Count);
        }
        
        public void ShipNameClearFilter()
        {
            this.GridPage.NavigateTo();
            this.CreateNewItemInDb();

            // Filter by GUID and we know we wait the grid to be empty
            this.GridPage.Grid.Filter(GridColumns.ShipName, Enums.FilterOperator.StartsWith, Guid.NewGuid().ToString());
            this.WaitForGridToLoad(0, this.GridPage.Grid);
            // Remove all filters and we expect that the grid will contain at least 1 item.
            this.GridPage.Grid.RemoveFilters();
            WaitForGridToLoadAtLeast(1, this.GridPage.Grid);
        }
    }
}

File: java/WebAutomation-Series/src/test/java/kendogrid/reuse/ShipNameColumnAsserter.java
package kendogrid.reuse;

import kendogrid.components.FilterOperator;
import kendogrid.components.GridFilter;
import kendogrid.components.GridItem;
import org.openqa.selenium.WebDriver;
import org.testng.Assert;

import java.util.List;
import java.util.UUID;

public class ShipNameColumnAsserter extends GridColumnAsserter {
    public ShipNameColumnAsserter(IGridPage gridPage, WebDriver driver) {
        super(gridPage, driver);
    }

    public void shipNameEqualToFilter() throws Exception {
        getGridPage().navigateTo();
        var newItem = createNewItemInDb();

        getGridPage().getGrid().filter(GridColumns.SHIP_NAME, FilterOperator.EQUAL_TO, newItem.getShipName());
        waitForGridToLoad(1, getGridPage().getGrid());
        List<GridItem> items = getGridPage().getGrid().getItems();

        Assert.assertEquals(items.stream().count(), 1);
    }

    public void shipNameContainsFilter() throws Exception {
        getGridPage().navigateTo();
        var shipName = UUID.randomUUID().toString();

        // Remove first and last letter
        shipName = removeLastChar(removeFirstChar(shipName));
        var newItem = createNewItemInDb(shipName);

        getGridPage().getGrid().filter(GridColumns.SHIP_NAME, FilterOperator.CONTAINS, newItem.getShipName());
        waitForGridToLoad(1, getGridPage().getGrid());
        List<GridItem> items = getGridPage().getGrid().getItems();

        Assert.assertEquals(items.stream().count(), 1);
    }

    public void shipNameEndsWithFilter() throws Exception {
        getGridPage().navigateTo();

        // Remove first letter
        var shipName = UUID.randomUUID().toString();
        shipName = removeFirstChar(shipName);
        var newItem = createNewItemInDb(shipName);

        getGridPage().getGrid().filter(GridColumns.SHIP_NAME, FilterOperator.ENDS_WITH, newItem.getShipName());
        waitForGridToLoad(1, getGridPage().getGrid());
        List<GridItem> items = getGridPage().getGrid().getItems();

        Assert.assertEquals(items.stream().count(), 1);
    }

    public void shipNameStartsWithFilter() throws Exception {
        getGridPage().navigateTo();

        // Remove last letter
        var shipName = UUID.randomUUID().toString();
        shipName = removeFirstChar(shipName);
        var newItem = createNewItemInDb(shipName);

        getGridPage().getGrid().filter(GridColumns.SHIP_NAME, FilterOperator.STARTS_WITH, newItem.getShipName());
        waitForGridToLoad(1, getGridPage().getGrid());
        List<GridItem> items = getGridPage().getGrid().getItems();

        Assert.assertEquals(items.stream().count(), 1);
    }

    public void shipNameNotEqualToFilter() throws Exception {
        getGridPage().navigateTo();

        // Apply combined filter. First filter by ID and than by ship name (not equal filter).
        // After the first filter there is only one element when we apply the second we expect 0 elements.
        var newItem = createNewItemInDb();

        getGridPage().getGrid().filter(
                new GridFilter(GridColumns.SHIP_NAME, FilterOperator.NOT_EQUAL_TO, newItem.getShipName()),
                new GridFilter(GridColumns.ORDER_ID, FilterOperator.EQUAL_TO, newItem.getOrderId().toString()));
        waitForGridToLoad(0, getGridPage().getGrid());
        List<GridItem> items = getGridPage().getGrid().getItems();

        Assert.assertEquals(items.stream().count(), 0);
    }

    public void shipNameNotContainsFilter() throws Exception {
        getGridPage().navigateTo();

        // Remove first and last letter
        var shipName = UUID.randomUUID().toString();
        shipName = removeLastChar(removeFirstChar(shipName));
        var newItem = createNewItemInDb(shipName);

        // Apply combined filter. First filter by ID and than by ship name (not equal filter).
        // After the first filter there is only one element when we apply the second we expect 0 elements.
        getGridPage().getGrid().filter(
                new GridFilter(GridColumns.SHIP_NAME, FilterOperator.NOT_CONTAINS, newItem.getShipName()),
                new GridFilter(GridColumns.ORDER_ID, FilterOperator.EQUAL_TO, newItem.getOrderId().toString()));
        waitForGridToLoad(0, getGridPage().getGrid());
        List<GridItem> items = getGridPage().getGrid().getItems();

        Assert.assertEquals(items.stream().count(), 0);
    }

    public void shipNameClearFilter() throws Exception {
        getGridPage().navigateTo();
        createNewItemInDb();

        // Filter by GUID and we know we wait the grid to be empty
        getGridPage().getGrid().filter(GridColumns.SHIP_NAME, FilterOperator.STARTS_WITH, UUID.randomUUID().toString());
        waitForGridToLoad(0, getGridPage().getGrid());

        // Remove all filters and we expect that the grid will contain at least 1 item.
        getGridPage().getGrid().removeFilters();
        waitForGridToLoadAtLeast(1, getGridPage().getGrid());
    }

    private String removeFirstChar(String s) {
        return s.substring(1);
    }

    private String removeLastChar(String s) {
        return s.substring(0, s.length() - 1);
    }
}
